lib/chef/provider/user/mac.rb

```diff
@@ -0,0 +1,607 @@
+#
+# Author:: Ryan Cragun (<ryan@chef.io>)
+# Copyright:: Copyright (c) 2019, Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../../resource"
+require_relative "../../dsl/declare_resource"
+require_relative "../../mixin/shell_out"
+require_relative "../../mixin/which"
+require_relative "../user"
+require_relative "../../resource/user/mac_user"
+
+class Chef
+  class Provider
+    class User
+      # A macOS user provider that is compatible with default TCC restrictions
+      # in macOS >= 10.14. See resource/user/mac_user.rb for complete description
+      # of the mac_user resource and how it differs from the dscl resource used
+      # on previous platforms.
+      class MacUser < Chef::Provider::User
+        include Chef::Mixin::Which
+
+        provides :mac_user
+        provides :user, os: "darwin", platform_version: ">= 10.14"
+
+        attr_accessor :auth_authority, :guid, :user_plist
+
+        DSCL_PROPERTY_MAP = {
+            uid: "dsAttrTypeStandard:UniqueID",
+            guid: "dsAttrTypeStandard:GeneratedUID",
+            gid: "dsAttrTypeStandard:PrimaryGroupID",
+            home: "dsAttrTypeStandard:NFSHomeDirectory",
+            shell: "dsAttrTypeStandard:UserShell",
+            comment: "dsAttrTypeStandard:RealName",
+            password: "dsAttrTypeStandard:Password",
+            auth_authority: "dsAttrTypeStandard:AuthenticationAuthority",
+            shadow_hash: "dsAttrTypeNative:ShadowHashData",
+            group_members: "dsAttrTypeStandard:GroupMembers",
+        }.freeze
+
+        def define_resource_requirements
+          super
+
+          %w{dscl sysadminctl plutil dsimport}.each do |bin|
+            requirements.assert(:all_actions) do |a|
+              a.assertion { which(bin) }
+              a.failure_message(Chef::Exceptions::User, "Cannot find binary '#{bin}' on the system for #{new_resource}!")
+            end
+          end
```

Comment:
is this really something that you ever reasonably expect to fail?

Reply:
I haven't tested it on platforms that aren't 10.14, but I would expect this resource and provider to _mostly_ work on 10.12 -> 10.14, aside from the SecureToken bits. If people specify  `mac_user` on <= 10.11 some of those probably won't be available. I also have absolutely no idea what strange ways Apple is going to change things for 10.15 either, so I was kind of trying to future proof for that case. That said, I know there are some incompatibilities with this and the dev preview of 10.15 that I'm trying to work out, but I'm not there yet.

On the only targeted platform for this resource (10.14) those binaries should almost surely exist. I'm happy to remove the check if you think it's adding nothing.

Reply:
yeah defensive coding that won't ever get exercised is really just cognitive overhead for anyone reading the code.

i also have a YAGNI instinct now when it comes to future-proofing...